homeworks/week3/Duy-hw.py

```diff
@@ -0,0 +1,63 @@
+class Node:
+    def __init__(self, val):
+        self.val = val 
+        self.next = None
+class SinglyLinkedList:
+    def __init__(self):
+        self.head = Node(float('-inf')) #sentinel node
+    def add(self, index, e):
+        '''Add e to specific index. Do nothing if index is OOR or e is None'''
+        #Time Complexity: O(N), Space: O(1)
+        if not e: #e == None
+            return
+        prev = self.head 
+        cur = self.head.next
+        while cur and index > 0:
+            index -= 1
+            prev = cur
+            cur = cur.next 
+        if index > 0: #Out of range
+            return
+        else:  #Valid
+            prev.next = e
+            e.next = cur
+
+    def remove(self, e):
+        '''Remove the 1st occurence of e.val from list. Return True if found, else False'''
+        #Time Complexity: O(N), Space: O(1)
+        if not e:
+            return False 
+        cur = self.head.next
+        prev = self.head
```

Comment:
Same as the previous comment, you can try to use only 1 variable instead of two. The value of this exercise is it allows you to view a problem in a different ways.